concurrency/TreatmentManager.java

```diff
@@ -0,0 +1,622 @@
+        + " " + e.getMessage());
+    }
+
+    treatmentIndex++;
+
+    synchronized (treatments) {
+        if (treatmentIndex >= treatments.size()) {
+            /*
+             * We're done playing all of the data, repeat if necessary
+             */
+            if (repeatCount != -1) {
+            if (repeatCount == 0) {
+                if (Logger.logLevel >= Logger.LOG_MOREINFO) {
+                    Logger.println("done playing treatment "
+                    + treatment);
+                }
+                stopTreatment();
+                return null;    // all done
+            }
+            repeatCount--;
+            }
+            treatmentIndex = 0;
+        }
+    }
+
+    sampleRateConverter = null;
+    audioSource = getAudioSource();
+
+    if (audioSource == null) {
+        return null;
+    }
+
+    try {
+        linearData = audioSource.getLinearData(sampleTime);
+
+        if (linearData != null && sampleRateConverter != null) {
+        linearData = sampleRateConverter.resample(linearData);
+        }
+
+        return linearData;
+    } catch (IOException e) {
+        Logger.println("Can't read linear data for " + treatment
+        + " " + e.getMessage());
+    }
+    return null;
+    }
+
+    private AudioSource getAudioSource() {
+    synchronized (treatments) {
+        if (treatments.size() == 0) {
+        if (Logger.logLevel >= Logger.LOG_MOREINFO) {
+            Logger.println("Treatments list is empty for " + treatment);
+        }
+            return null;
+        }
+
+        AudioSource audioSource = treatments.get(treatmentIndex);
+
+        if (sampleRateConverter != null) {
+            return audioSource;
+        }
+
+        if (sampleRate != audioSource.getSampleRate() ||
+                channels != audioSource.getChannels()) {
+
+            try {
+                sampleRateConverter =
+                new SampleRateConverter("Treatment",
+                audioSource.getSampleRate(),
+                    audioSource.getChannels(),
+                sampleRate, channels);
+            } catch (IOException e) {
+            Logger.println("Can't resample treatment! "
+                + e.getMessage());
+            return null;
+            }
+        }
+        return audioSource;
+    }
+    }
+
+    /*
+     * Register to be notified when the treatment finishes.
+     */
+    public void addTreatmentDoneListener(TreatmentDoneListener listener) {
+    synchronized (treatmentDoneListeners) {
+        treatmentDoneListeners.add(listener);
+    }
+    }
+
+    public void removeTreatmentDoneListener(TreatmentDoneListener listener) {
+    synchronized (treatmentDoneListeners) {
+        treatmentDoneListeners.remove(listener);
+    }
+    }
+
+    public void stopTreatment() {
+    stopTreatment(true);
+    }
+
+    public void stopTreatment(boolean notify) {
+    if (isStopped) {
+        return;
+    }
+
+    isStopped = true;
+
+    for (AudioSource audioSource : treatments) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 This line may not be thread-safe: **339**. At this line the following field is not protected by synchronization: `treatments`. Your code typically protects this field with synchronization on object: `treatments`.

**Fix**  
 Consider protecting the above field with synchronization on object: `treatments`. For example, like at lines: **527**, **344**, and **578**.

**More info**  
 The following lines are where the above field is already protected by synchronization on the above object: **238**, **281**, **288**, **344**, **386**, **397**, **527**, and **578**.